### ISSUE TITLE
Cross-Protocol AdBlock and Tracking Protection (QUIC over UDP) 

### ISSUE THREAD

**Issue by lukemulks** [brave/browser-laptop#6843]
- Did you search for similar issues before submitting this one?
Yes

- Describe the issue you encountered:
Long term follow-up to issue #6831 

- Platform (Win7, 8, 10? macOS? Linux distro?):
All 

- Brave Version (revision SHA):
NA

**Original Issue:** 
It was discovered that Google was using QUIC over UDP for ad and tracking purposes, which bypassed traditional TCP list filtering.  

**Short term fix:** 
#6831 (blocking QUIC over UDP protocol) 

**Longer term:** 
**_Option I:_** Apply existing AdBlocking & Tracking Protection lists to QUIC over UDP protocol. 
- This may prove difficult, if QUIC over UDP ad/tracking requests are formed differently than TCP. 
Need to investigate and compare QUIC vs TCP requests to determine if TCP-style blocklists will transition cleanly to QUIC, or if separate blocklists are required for atypical protocols. 

**_Option II:_** Keep QUIC over UDP protocol blocked, and perform exception-based whitelisting for services that don't prove to be ad/tracking related. 
- May become a manual pain longer-term, but might not. Depends longer-term protocol adoption by vendors and products. One indicator: https://datatracker.ietf.org/wg/quic/documents/ A QUIC working group was started in Oct. 2016 to standardize QUIC. 

**QUIC standardization**
 Milestones from the QUIC Working Group: https://datatracker.ietf.org/wg/quic/charter/

```
May 2019    Multipath extension document to IESG
Nov 2018    QUIC Applicability and Manageability Statement to IESG
Nov 2018    HTTP/2 mapping document to IESG
Mar 2018    TLS 1.3 Mapping document to IESG
Mar 2018    Loss detection and Congestion Control document to IESG
Mar 2018    Core Protocol document to IESG
Nov 2017    Working group adoption of Multipath extension document
Feb 2017    Working group adoption of QUIC Applicability and Manageability Statement
Feb 2017    Working group adoption of HTTP/2 mapping document
Feb 2017    Working group adoption of TLS 1.3 mapping document
Feb 2017    Working group adoption of Loss detection and Congestion Control document
Feb 2017    Working group adoption of Core Protocol document
```

**Additional info on QUIC:** 
QUIC FAQ: https://docs.google.com/document/d/1lmL9EF6qKrk7gbazY8bIdvq3Pno2Xj_l_YShP40GLQE/mobilebasic

QUIC: Design Document & Spec Rationale: 
https://docs.google.com/document/d/1RNHkx_VvKWyWg6Lr8SZ-saqsQx7rFV-ev2jRFUoVD34/mobilebasic

QUIC_Optimizing network performance - Chrome DevSummit 2013
[QUIC_Optimizing network performance - Chrome DevSummit 2013.pdf](https://github.com/brave/browser-laptop/files/728433/QUIC_Optimizing.network.performance.-.Chrome.DevSummit.2013.pdf)

**Comment by lukemulks** [comment]
High-level Chromium reference on how QUIC protocol is selected in requests
https://chromium.googlesource.com/chromium/src/+/lkgr/net/docs/life-of-a-url-request.md#QUIC

*QUIC*
>`QUIC` works quite a bit differently from `HTTP/2`. Servers advertise `QUIC` support with an `Alternate-Protocol` HTTP header in their responses. `HttpServerProperties` then tracks servers that have advertised `QUIC` support.
>
>When a new request comes in to `HttpStreamFactoryImpl` for a connection to a server that has advertised `QUIC` support in the past, it will create a second `HttpStreamFactoryImpl::Job` for `QUIC`, which returns an `QuicHttpStream` on success. The two Jobs (One for `QUIC`, one for all versions of `HTTP`) will be raced against each other, and whichever successfully creates an `HttpStream` first will be used.
>
>As with `HTTP/2`, once a `QUIC` connection is established, it will be shared with other `Jobs` connecting to the same server, and future `Jobs` will just reuse the existing `QUIC` session.

*Prioritization*
>`URLRequests` are assigned a priority on creation. It only comes into play in a couple places:
>
> *---* The `ResourceScheduler` lives outside `net/`, and in some cases, delays starting low priority requests on a per-tab basis.
>
> *---* `DNS lookups` are initiated based on the highest priority request for a lookup.
>
> *---* Socket pools hand out and create sockets based on prioritization. However, when a socket becomes idle, it will be assigned to the highest priority request for the server its connected to, even if there‘s a higher priority request to another server that’s waiting on a free socket slot.
>
> *---* `HTTP/2` and `QUIC` both support sending priorities over-the-wire.
>
>At the socket pool layer, sockets are only assigned to socket requests once the socket is connected and SSL is negotiated, if needed. This is done so that if a higher priority request for a group reaches the socket pool before a connection is established, the first usable connection goes to the highest priority socket request.

**Comment by lukemulks** [comment]
**For a QUIC Google Filter List**  @diracdeltas  @bridiver @mrose17 

- Below are QUIC sessions pulled from a traffic capture from Chrome on Android (Nexus 9 Tablet) over Wifi. 

- Each domain listed below contains one ad-related domain that's wildcarded among the domains in the policy pool. 

- Some of the request domains appear redundant, but aren't. Each is used this way to either cachebust the request from the domain-level, or resist domain level blocking. 

```t=  5550 [st=     0] +QUIC_SESSION  [dt=145463+]
                     --> cert_verify_flags = 6
                     --> host = "static.doubleclick.net"
                     --> port = 443
t=  6190 [st=   640]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.doubleclick.net","*.2mdn.net","*.au.doubleclick.net","*.cc-dt.com","*.de.doubleclick.net","*.doubleclick.com","*.fls.doubleclick.net","*.fr.doubleclick.net","*.jp.doubleclick.net","*.uk.doubleclick.net","ad.mo.doubleclick.net","doubleclick.com","doubleclick.net"]
t=  6190 [st=   640]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t=  7525 [st=     0] +QUIC_SESSION  [dt=143488+]
                     --> cert_verify_flags = 6
                     --> host = "stats.g.doubleclick.net"
                     --> port = 443
```t=  7747 [st=   222]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.g.doubleclick.net","*.googleadservices.com","*.googlesyndication.com","*.googletagservices.com","*.invitemedia.com","g.doubleclick.net","googleadservices.com","googlesyndication.com","googletagservices.com","media.admob.com"]
t=  7747 [st=   222]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```


>*-------------------------------------------------------*
>*-------------------------------------------------------*

t=  5543 [st=     0] +QUIC_SESSION  [dt=145470+]
                     --> cert_verify_flags = 6
                     --> host = "s.ytimg.com"
                     --> port = 443
```t=  5987 [st=   444]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.google.com","*.android.com","*.appengine.google.com","*.cloud.google.com","*.gcp.gvt2.com","*.google-analytics.com","*.google.ca","*.google.cl","*.google.co.in","*.google.co.jp","*.google.co.uk","*.google.com.ar","*.google.com.au","*.google.com.br","*.google.com.co","*.google.com.mx","*.google.com.tr","*.google.com.vn","*.google.de","*.google.es","*.google.fr","*.google.hu","*.google.it","*.google.nl","*.google.pl","*.google.pt","*.googleadapis.com","*.googleapis.cn","*.googlecommerce.com","*.googlevideo.com","*.gstatic.cn","*.gstatic.com","*.gvt1.com","*.gvt2.com","*.metric.gstatic.com","*.urchin.com","*.url.google.com","*.youtube-nocookie.com","*.youtube.com","*.youtubeeducation.com","*.ytimg.com","android.clients.google.com","android.com","developer.android.google.cn","g.co","goo.gl","google-analytics.com","google.com","googlecommerce.com","urchin.com","www.goo.gl","youtu.be","youtube.com","youtubeeducation.com"]
t=  5988 [st=   445]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t=  8611 [st=     0] +QUIC_SESSION  [dt=142402+]
                     --> cert_verify_flags = 6
                     --> host = "www.gstatic.com"
                     --> port = 443
```t=  8909 [st=   298]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.google.com","*.android.com","*.appengine.google.com","*.cloud.google.com","*.gcp.gvt2.com","*.google-analytics.com","*.google.ca","*.google.cl","*.google.co.in","*.google.co.jp","*.google.co.uk","*.google.com.ar","*.google.com.au","*.google.com.br","*.google.com.co","*.google.com.mx","*.google.com.tr","*.google.com.vn","*.google.de","*.google.es","*.google.fr","*.google.hu","*.google.it","*.google.nl","*.google.pl","*.google.pt","*.googleadapis.com","*.googleapis.cn","*.googlecommerce.com","*.googlevideo.com","*.gstatic.cn","*.gstatic.com","*.gvt1.com","*.gvt2.com","*.metric.gstatic.com","*.urchin.com","*.url.google.com","*.youtube-nocookie.com","*.youtube.com","*.youtubeeducation.com","*.ytimg.com","android.clients.google.com","android.com","developer.android.google.cn","g.co","goo.gl","google-analytics.com","google.com","googlecommerce.com","urchin.com","www.goo.gl","youtu.be","youtube.com","youtubeeducation.com"]
t=  8909 [st=   298]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 10892 [st=     0] +QUIC_SESSION  [dt=140121+]
                     --> cert_verify_flags = 6
                     --> host = "googleads.g.doubleclick.net"
                     --> port = 443
```t= 10983 [st=    91]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.g.doubleclick.net","*.googleadservices.com","*.googlesyndication.com","*.googletagservices.com","*.invitemedia.com","g.doubleclick.net","googleadservices.com","googlesyndication.com","googletagservices.com","media.admob.com"]
t= 10983 [st=    91]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```
 
>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 17164 [st=     0] +QUIC_SESSION  [dt=133849+]
                     --> cert_verify_flags = 6
                     --> host = "www.youtube.com"
                     --> port = 443
```t= 17585 [st=   421]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.google.com","*.android.com","*.appengine.google.com","*.cloud.google.com","*.gcp.gvt2.com","*.google-analytics.com","*.google.ca","*.google.cl","*.google.co.in","*.google.co.jp","*.google.co.uk","*.google.com.ar","*.google.com.au","*.google.com.br","*.google.com.co","*.google.com.mx","*.google.com.tr","*.google.com.vn","*.google.de","*.google.es","*.google.fr","*.google.hu","*.google.it","*.google.nl","*.google.pl","*.google.pt","*.googleadapis.com","*.googleapis.cn","*.googlecommerce.com","*.googlevideo.com","*.gstatic.cn","*.gstatic.com","*.gvt1.com","*.gvt2.com","*.metric.gstatic.com","*.urchin.com","*.url.google.com","*.youtube-nocookie.com","*.youtube.com","*.youtubeeducation.com","*.ytimg.com","android.clients.google.com","android.com","developer.android.google.cn","g.co","goo.gl","google-analytics.com","google.com","googlecommerce.com","urchin.com","www.goo.gl","youtu.be","youtube.com","youtubeeducation.com"]
t= 17585 [st=   421]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 19756 [st=     0] +QUIC_SESSION  [dt=131257+]
                     --> cert_verify_flags = 6
                     --> host = "www.google-analytics.com"
                     --> port = 443
```t= 20139 [st=   383]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.google.com","*.android.com","*.appengine.google.com","*.cloud.google.com","*.gcp.gvt2.com","*.google-analytics.com","*.google.ca","*.google.cl","*.google.co.in","*.google.co.jp","*.google.co.uk","*.google.com.ar","*.google.com.au","*.google.com.br","*.google.com.co","*.google.com.mx","*.google.com.tr","*.google.com.vn","*.google.de","*.google.es","*.google.fr","*.google.hu","*.google.it","*.google.nl","*.google.pl","*.google.pt","*.googleadapis.com","*.googleapis.cn","*.googlecommerce.com","*.googlevideo.com","*.gstatic.cn","*.gstatic.com","*.gvt1.com","*.gvt2.com","*.metric.gstatic.com","*.urchin.com","*.url.google.com","*.youtube-nocookie.com","*.youtube.com","*.youtubeeducation.com","*.ytimg.com","android.clients.google.com","android.com","developer.android.google.cn","g.co","goo.gl","google-analytics.com","google.com","googlecommerce.com","urchin.com","www.goo.gl","youtu.be","youtube.com","youtubeeducation.com"]
t= 20140 [st=   384]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```
 
>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 20043 [st=     0] +QUIC_SESSION  [dt=130970+]
                     --> cert_verify_flags = 6
                     --> host = "fonts.googleapis.com"
                     --> port = 443
```t= 20447 [st=   404]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.googleapis.com","*.clients6.google.com","*.cloudendpointsapis.com","cloudendpointsapis.com","googleapis.com"]
t= 20447 [st=   404]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 29426 [st=     0] +QUIC_SESSION  [dt=121587+]
                     --> cert_verify_flags = 6
                     --> host = "securepubads.g.doubleclick.net"
                     --> port = 443
```t= 30000 [st=   574]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.g.doubleclick.net","*.googleadservices.com","*.googlesyndication.com","*.googletagservices.com","*.invitemedia.com","g.doubleclick.net","googleadservices.com","googlesyndication.com","googletagservices.com","media.admob.com"]
t= 30000 [st=   574]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 31294 [st=     0] +QUIC_SESSION  [dt=119719+]
                     --> cert_verify_flags = 6
                     --> host = "fonts.gstatic.com"
                     --> port = 443
```t= 31546 [st=   252]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.google.com","*.android.com","*.appengine.google.com","*.cloud.google.com","*.gcp.gvt2.com","*.google-analytics.com","*.google.ca","*.google.cl","*.google.co.in","*.google.co.jp","*.google.co.uk","*.google.com.ar","*.google.com.au","*.google.com.br","*.google.com.co","*.google.com.mx","*.google.com.tr","*.google.com.vn","*.google.de","*.google.es","*.google.fr","*.google.hu","*.google.it","*.google.nl","*.google.pl","*.google.pt","*.googleadapis.com","*.googleapis.cn","*.googlecommerce.com","*.googlevideo.com","*.gstatic.cn","*.gstatic.com","*.gvt1.com","*.gvt2.com","*.metric.gstatic.com","*.urchin.com","*.url.google.com","*.youtube-nocookie.com","*.youtube.com","*.youtubeeducation.com","*.ytimg.com","android.clients.google.com","android.com","developer.android.google.cn","g.co","goo.gl","google-analytics.com","google.com","googlecommerce.com","urchin.com","www.goo.gl","youtu.be","youtube.com","youtubeeducation.com"]
t= 31547 [st=   253]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 34353 [st=     0] +QUIC_SESSION  [dt=116660+]
                     --> cert_verify_flags = 6
                     --> host = "www.google.com"
                     --> port = 443
```                        --> ct_compliance_status = "COMPLIES_VIA_SCTS"
t= 34512 [st=   159]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.google.com","*.android.com","*.appengine.google.com","*.cloud.google.com","*.gcp.gvt2.com","*.google-analytics.com","*.google.ca","*.google.cl","*.google.co.in","*.google.co.jp","*.google.co.uk","*.google.com.ar","*.google.com.au","*.google.com.br","*.google.com.co","*.google.com.mx","*.google.com.tr","*.google.com.vn","*.google.de","*.google.es","*.google.fr","*.google.hu","*.google.it","*.google.nl","*.google.pl","*.google.pt","*.googleadapis.com","*.googleapis.cn","*.googlecommerce.com","*.googlevideo.com","*.gstatic.cn","*.gstatic.com","*.gvt1.com","*.gvt2.com","*.metric.gstatic.com","*.urchin.com","*.url.google.com","*.youtube-nocookie.com","*.youtube.com","*.youtubeeducation.com","*.ytimg.com","android.clients.google.com","android.com","developer.android.google.cn","g.co","goo.gl","google-analytics.com","google.com","googlecommerce.com","urchin.com","www.goo.gl","youtu.be","youtube.com","youtubeeducation.com"]
t= 34512 [st=   159]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 34392 [st=     0] +QUIC_SESSION  [dt=116621+]
                     --> cert_verify_flags = 6
                     --> host = "pagead2.googlesyndication.com"
                     --> port = 443
```t= 34630 [st=   238]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.googleusercontent.com","*.apps.googleusercontent.com","*.appspot.com.storage.googleapis.com","*.blogspot.com","*.bp.blogspot.com","*.commondatastorage.googleapis.com","*.content-storage-download.googleapis.com","*.content-storage-upload.googleapis.com","*.content-storage.googleapis.com","*.doubleclickusercontent.com","*.ggpht.com","*.googledrive.com","*.googlesyndication.com","*.googleweblight.com","*.safenup.googleusercontent.com","*.sandbox.googleusercontent.com","*.storage-download.googleapis.com","*.storage-upload.googleapis.com","*.storage.googleapis.com","blogspot.com","bp.blogspot.com","commondatastorage.googleapis.com","doubleclickusercontent.com","ggpht.com","googledrive.com","googleusercontent.com","googleweblight.com","static.panoramio.com.storage.googleapis.com","storage.googleapis.com","unfiltered.news"]
t= 34631 [st=   239]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 34398 [st=     0] +QUIC_SESSION  [dt=116615+]
                     --> cert_verify_flags = 6
                     --> host = "tpc.googlesyndication.com"
                     --> port = 443
```t= 34559 [st=   161]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.googleusercontent.com","*.apps.googleusercontent.com","*.appspot.com.storage.googleapis.com","*.blogspot.com","*.bp.blogspot.com","*.commondatastorage.googleapis.com","*.content-storage-download.googleapis.com","*.content-storage-upload.googleapis.com","*.content-storage.googleapis.com","*.doubleclickusercontent.com","*.ggpht.com","*.googledrive.com","*.googlesyndication.com","*.googleweblight.com","*.safenup.googleusercontent.com","*.sandbox.googleusercontent.com","*.storage-download.googleapis.com","*.storage-upload.googleapis.com","*.storage.googleapis.com","blogspot.com","bp.blogspot.com","commondatastorage.googleapis.com","doubleclickusercontent.com","ggpht.com","googledrive.com","googleusercontent.com","googleweblight.com","static.panoramio.com.storage.googleapis.com","storage.googleapis.com","unfiltered.news"]
t= 34559 [st=   161]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 46794 [st=     0] +QUIC_SESSION  [dt=104219+]
                     --> cert_verify_flags = 6
                     --> host = "ad.doubleclick.net"
                     --> port = 443
```t= 46828 [st=    34]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.doubleclick.net","*.2mdn.net","*.au.doubleclick.net","*.cc-dt.com","*.de.doubleclick.net","*.doubleclick.com","*.fls.doubleclick.net","*.fr.doubleclick.net","*.jp.doubleclick.net","*.uk.doubleclick.net","ad.mo.doubleclick.net","doubleclick.com","doubleclick.net"]
t= 46828 [st=    34]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 48351 [st=     0] +QUIC_SESSION  [dt=102662+]
                     --> cert_verify_flags = 6
                     --> host = "s0.2mdn.net"
                     --> port = 443
```t= 48418 [st=    67]    QUIC_SESSION_CERTIFICATE_VERIFIED
                        --> subjects = ["*.doubleclick.net","*.2mdn.net","*.au.doubleclick.net","*.cc-dt.com","*.de.doubleclick.net","*.doubleclick.com","*.fls.doubleclick.net","*.fr.doubleclick.net","*.jp.doubleclick.net","*.uk.doubleclick.net","ad.mo.doubleclick.net","doubleclick.com","doubleclick.net"]
t= 48418 [st=    67]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*

t= 54546 [st=    0] +QUIC_SESSION  [dt=96467+]
                    --> cert_verify_flags = 6
                    --> host = "googleads4.g.doubleclick.net"
                    --> port = 443
```t= 54676 [st=  130]    QUIC_SESSION_CERTIFICATE_VERIFIED
                       --> subjects = ["*.g.doubleclick.net","*.googleadservices.com","*.googlesyndication.com","*.googletagservices.com","*.invitemedia.com","g.doubleclick.net","googleadservices.com","googlesyndication.com","googletagservices.com","media.admob.com"]
t= 54676 [st=  130]    QUIC_SESSION_CRYPTO_HANDSHAKE_MESSAGE_SENT```

>*-------------------------------------------------------*
>*-------------------------------------------------------*
